Src/Processor/Processor.cpp

```diff
@@ -131,6 +109,144 @@ void CProcessor::MonitorNewScores()
 #endif
 }
 
+void CProcessor::ProcessAllUsers(bool reProcess, u32 numThreads)
+{
+	CThreadPool threadPool{numThreads};
+	std::vector<std::shared_ptr<CDatabaseConnection>> databaseConnections;
```

Comment:
In this case I'd argue for leaving the `shared_ptr` in. `CUpdateBatch` _should_ accept and store a `shared_ptr` to make it harder to use incorrectly (i.e. to leave its reference to a `CDatabaseConnection` dangling), regardless of whether that's required in this specific use case or not. The performance overhead is basically zero, so a safe API should be preferred.

That said, `CDatabaseConnection` needs its copy constructor deleted and a move constructor implemented, because using it by-value most definitely is broken at the moment.

Reply:
I guess `databaseConnections` exclusively could store raw pointers instead of `shared_ptr` while leaving everything else equal, but that'd invite an error down the line and, again, really is not worth the negligible performance gain of incrementing and decrementing `numThreads` ref-counts.

Reply:
> That said, CDatabaseConnection needs its copy constructor deleted and a move constructor implemented, because using it by-value most definitely is broken at the moment.
> I guess databaseConnections exclusively could store raw pointers instead of shared_ptr while leaving everything else equal, [...]

I see. Still, it could still be a `std::vector<std::unique_ptr<...>>` instead. `ProcessSingleUser` has no business receiving anything more than a raw pointer, it just wants a view.

> [...] but that'd invite an error down the line and, again, really is not worth the negligible performance gain of not incrementing and decrementing numThreads ref-counts.

{in,de}crementing ref-counts atomically, and copying `std::shared_ptr`s for every single user when the simplest form of RAII seems like a waste of performance (it really tends to be high), but you're right I don't know the performance characteristics of the program.

Reply:
`unique_ptr` would work just as well in this case, yeah, but it'd require returning raw pointers from the helper functions creating the database connections, which in turn may lead to unsafe code at some point in the future. Really, the argument is that SQL Connections are _inherently_ so much more expensive to create/drop than an atomic inc/dec that it _never_ would be interesting from a performance standpoint to optimise at that point.

You are right, that the connections are unnecessarily passed as `shared_ptr` to `ProcessSingleUser`, but references can just be passed instead in this case rather than removing the `shared_ptr`s altogether. That said, this change would not affect performance in any meaningful way since the amount of computation _per user_ involves churning through thousands of database records and massaging them (in-fact super-linearly in cost).

I fully agree that removal of `shared_ptr` in favour of `unique_ptr`/raw/ref in inner hot loops would be an important thing to do.